### ISSUE TITLE
blocks_in_conditions suggests a fix which breaks code due to lifetime extension rules

### ISSUE THREAD

**Issue by polina4096** [rust-lang/rust-clippy#15112]
### Summary

For example, if a mutex is locked and its value is matched in the match condition, the lock is actually held for the duration of the entire match and is not dropped early due to lifetime extension rules.

This could easily become problematic because usually it's expected that the futures are `Send` and due to the mutex guard being not `Send` holding it across await points also makes the futures not `Send`.

As such, one can wrap the condition in a block to narrow the scope of the mutex guard. However clippy gets angry and suggests to remove the block which in turn extends the scope and breaks the code.

```rust
tokio::spawn(async {
  let state = Mutex::new(10i32);

  // Clippy gets angry about this block.
  // However code does not cpile if we remove it.
  match { *state.lock().unwrap() } {
    1 .. 20 => {
      // Futures which hold a lock across await points are not `Send`,
      // so if the block above is removed, it's no longer possible to spawn
      // it because `tokio::spawn` requires futures to be `Send`.
      tokio::time::sleep(std::time::Duration::from_secs(1)).await;
    }

    _ => {}
  };
});
```

### Reproducer

I tried this code:

```rust
<code>
```

I saw this happen:

```
<output>
```

I expected to see this happen:

### Version

```text

```

**Comment by profetia** [comment]
I think this lint should not trigger when there is any non-trivial drop happens in the block
